app/models/user.rb

```diff
@@ -49,7 +49,7 @@ class User < ApplicationRecord
          :recoverable, :trackable, :validatable
   include DeviseTokenAuth::Concerns::User
 
-  has_many :preferences, dependent: :destroy
+  has_many :preferences, dependent: :destroy, before_add: :check_preferences_limit
```

Comment:
Adding `before_add` on `User` model is one option. Other option could be to add `validate :check_preferences_limit, on: :create` on `Preference` model.
The difference is where and how the preference limit is applied:
- `before_add` on `User`: Applies the limit **only** when preferences are added through `user.preferences`. It will not be applied if preferences are created **directly** in Preference.
- `validate` in `Preference`: Applies the limit **consistently**, regardless of whether the preferences are added through User or directly in Preference.

For your code, I see no problem on have in it on User, but keep in mind this difference for future, because it will depend on the context.